src/sdk/namespace_fs.js

```diff
@@ -492,23 +494,94 @@ class NamespaceFS {
     ///////////////////
 
     async upload_object(params, object_sdk) {
-        try {
-            const fs_account_config = this.set_cur_fs_account_config(object_sdk);
-            await this._load_bucket(params, fs_account_config);
+        const fs_account_config = this.set_cur_fs_account_config(object_sdk);
+        await this._load_bucket(params, fs_account_config);
 
-            const file_path = this._get_file_path(params);
-            const upload_id = uuidv4();
-            const upload_path = path.join(this.bucket_path, config.NSFS_TEMP_DIR_NAME, 'uploads', upload_id);
-            // console.log('NamespaceFS.upload_object:', upload_path, '->', file_path);
+        const file_path = this._get_file_path(params);
+        // Uploading to a temp file then we will rename it. 
+        const upload_id = uuidv4();
+        const upload_path = path.join(this.bucket_path, config.NSFS_TEMP_DIR_NAME, 'uploads', upload_id);
+        // console.log('NamespaceFS.upload_object:', upload_path, '->', file_path);
+        await Promise.all([this._make_path_dirs(file_path, fs_account_config), this._make_path_dirs(upload_path, fs_account_config)]);
+        if (params.copy_source) {
+            const source_file_path = path.join(this.bucket_path, params.copy_source.key);
+            try {
+                const same_inode = await this._is_same_inode(fs_account_config, source_file_path, file_path);
+                if (same_inode) return same_inode;
```

Comment:
A bit confusing to understand.
I would expect `_is_same_inode` to return a Boolean.
Instead of either returns an Object or `false`.
A suggestion is to make the function return Object or nothing and rename it in order to be more understandable.

Reply:
We will need this logic somewhere, as when we are deciding not to do anything the copy should return the etag. 
and since we have the etag as we did `stat` inside `_is_same_inode` we can return it. 
so, if `same_inode` returns false we will try and link, and if it returns the etag then we are set and just return it.